
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Modules &#8212; PacBio Data Processing 0.20.2 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css" />
    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="How to contribute" href="../contributing.html" />
    <link rel="prev" title="Design notes" href="design.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="modules">
<h1>Modules<a class="headerlink" href="#modules" title="Permalink to this headline">¶</a></h1>
<section id="module-pacbio_data_processing.sm_analysis">
<span id="sm-analysis"></span><h2>sm_analysis<a class="headerlink" href="#module-pacbio_data_processing.sm_analysis" title="Permalink to this headline">¶</a></h2>
<p>This module contains the high level functions necessary to run
the ‘Single Molecule Analysis’ on an input BAM file.</p>
<dl class="py function">
<dt class="sig sig-object py" id="pacbio_data_processing.sm_analysis.add_to_own_output">
<span class="sig-prename descclassname"><span class="pre">pacbio_data_processing.sm_analysis.</span></span><span class="sig-name descname"><span class="pre">add_to_own_output</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gffs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">own_output_file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">modification_types</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pacbio_data_processing/sm_analysis.html#add_to_own_output"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pacbio_data_processing.sm_analysis.add_to_own_output" title="Permalink to this definition">¶</a></dt>
<dd><p>From a set of .gff files, a csv file (delimiter=”,”) is saved
with the following columns:</p>
<blockquote>
<div><ul class="simple">
<li><p>mol id: taken each gff file (e.g. ‘a.b.c.gff’ -&gt; mol id: ‘b’)</p></li>
<li><p>modtype: column number 3 (idx: 2) of the gffs (feature type)</p></li>
<li><p>GATC position: column number column number 5 (idx: 4) of the
gffs which corresponds to the ‘end coordinate of the feature’
in the GFF3 standard</p></li>
<li><p>score of the feature: column number 6 (idx: 5); floating point
(Phred-transformed pvalue that a kinetic deviation exists at
this position)</p></li>
<li><p>strand: strand of the feature. It can be +, - with obvious
meanings. It can also be ? (meaning unknown) or . (for non
stranded features)</p></li>
</ul>
</div></blockquote>
<p>There are more columns, but they are nor fixed in number. They
correspond to the values given in the ‘attributes’ column of the
gffs (col 9, idx 8).
For example, given the following attributes column:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">coverage</span><span class="o">=</span><span class="mi">134</span><span class="p">;</span><span class="n">context</span><span class="o">=</span><span class="n">TCA</span><span class="o">...</span><span class="p">;</span><span class="n">IPDRatio</span><span class="o">=</span><span class="mf">3.91</span><span class="p">;</span><span class="n">identificationQv</span><span class="o">=</span><span class="mi">228</span>
</pre></div>
</div>
<p>we would get the following ‘extra’ columns:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">134</span><span class="p">,</span><span class="n">TCA</span><span class="o">...</span><span class="p">,</span><span class="mf">3.91</span><span class="p">,</span><span class="mi">228</span>
</pre></div>
</div>
<p>and this is exactly what happens with the m6A modification type.</p>
<p>All the lines starting by ‘#’ in the gff files are ignored.
The format of the gff file is GFF3:
<a class="reference external" href="https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md">https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md</a></p>
<p>The value of identificationQV is a a phred transformed probability of
having a detection. See eq. (8) in [1]</p>
<p>[1]: “Detection and Identification of Base Modifications with Single
Molecule Real-Time Sequencing Data”</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pacbio_data_processing.sm_analysis.generate_CCS_file">
<span class="sig-prename descclassname"><span class="pre">pacbio_data_processing.sm_analysis.</span></span><span class="sig-name descname"><span class="pre">generate_CCS_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ccs</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><a class="reference internal" href="../pacbio_data_processing.html#pacbio_data_processing.external.CCS" title="pacbio_data_processing.external.CCS"><span class="pre">pacbio_data_processing.external.CCS</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_bam</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><span class="pre">pathlib.Path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ccs_bam_file</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><span class="pre">pathlib.Path</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">pathlib.Path</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pacbio_data_processing/sm_analysis.html#generate_CCS_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pacbio_data_processing.sm_analysis.generate_CCS_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Idempotent computation of the Circular Consensus Sequence (CCS)
version of the passed in <code class="docutils literal notranslate"><span class="pre">in_bam</span></code> file done with passed-in <code class="docutils literal notranslate"><span class="pre">ccs</span></code>
object.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the CCS bam file, if it is there, or <code class="docutils literal notranslate"><span class="pre">None</span></code> if if could
not be computed (yet).</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pacbio_data_processing.sm_analysis.main_cl">
<span class="sig-prename descclassname"><span class="pre">pacbio_data_processing.sm_analysis.</span></span><span class="sig-name descname"><span class="pre">main_cl</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="../_modules/pacbio_data_processing/sm_analysis.html#main_cl"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pacbio_data_processing.sm_analysis.main_cl" title="Permalink to this definition">¶</a></dt>
<dd><p>Entry point for <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> executable.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pacbio_data_processing.sm_analysis.map_molecules_with_highest_sim_ratio">
<span class="sig-prename descclassname"><span class="pre">pacbio_data_processing.sm_analysis.</span></span><span class="sig-name descname"><span class="pre">map_molecules_with_highest_sim_ratio</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bam_file_name</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">pathlib.Path</span><span class="p"><span class="pre">,</span> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span> </span><a class="reference internal" href="../pacbio_data_processing.html#pacbio_data_processing.bam_utils.Molecule" title="pacbio_data_processing.bam_utils.Molecule"><span class="pre">pacbio_data_processing.bam_utils.Molecule</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/pacbio_data_processing/sm_analysis.html#map_molecules_with_highest_sim_ratio"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pacbio_data_processing.sm_analysis.map_molecules_with_highest_sim_ratio" title="Permalink to this definition">¶</a></dt>
<dd><p>Given the path to a bam file, it returns a dictionary, whose
keys are mol ids (ints) and the values are the corresponding
Molecules.
If multiple lines in the given BAM file share the mol id, only
the first line found with the highest similarity ratio (computed from
the cigar) is chosen: if multiple lines share the molecule ID and
the highest similarity ratio (say, 1), ONLY the first one is taken,
irrespective of other factors.</p>
</dd></dl>

</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">PacBio Data Processing</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../readme.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quickstart.html">Quickstart for the impatient</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quickstart_9steps.html">Quickstart on a cluster</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage/index.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../indepth/index.html">Guides</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Developer’s documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules.html">pacbio_data_processing</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="index.html">Developer’s documentation</a><ul>
      <li>Previous: <a href="design.html" title="previous chapter">Design notes</a></li>
      <li>Next: <a href="../contributing.html" title="next chapter">How to contribute</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021-2022, David Palao and David Velázquez.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/devel/modules.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>